Initial https support for wordpress

NitinAtIntel opened the pull request:
Signed-off-by: NitinAtIntel <nitin.tekchandani@intel.com>

suresh-srinivas commented:
This looks great. Thanks for starting this.

I would suggest using the same Nginx for both the base and the opt. And use a variable to enable/disable the OpenSSL crypto engine. See the following
https://github.com/intel/crypto-reference-stack/blob/main/wordpress/centos8/Dockerfile

Would also suggest providing the NGINX_CIPHER as a variable.

And for us to be able to compare this to the wp_base and wp_opt, you can also have the NGINX_HTTP variable which controls the server type like in the reference stack.

suresh-srinivas commented on containers/wordpress/wp_base/https/Dockerfile:
```diff
@@ -0,0 +1,83 @@
+    libpcre3-dev
+
+
+# Build/Install openssl v1_1_1f
```
Should we be using OpenSSL_1_1_1j

NitinAtIntel replied:
This was based on @uttampawar's internal reference but if no need to stay with f we can certainly move to j.

suresh-srinivas commented on containers/wordpress/wp_base/https/Dockerfile:
```diff
@@ -0,0 +1,83 @@
+RUN openssl dhparam -out dhparam.pem 2048
+RUN chown -R ${USERNAME} /home/${USERNAME}/certificates
+
+# NGINX
```
Suggest to use the NGNIX Async in the base. https://github.com/intel/asynch_mode_nginx.git

NitinAtIntel replied:
This has been done in the latest commit.

suresh-srinivas commented on containers/wordpress/wp_base/https/files/nginx.conf.in:
```diff
@@ -0,0 +1,122 @@
+worker_processes  96; # Depends on #vcpus
```
It would be better to make this a variable. Also if we use NGNIX Async, do we need the nginx.conf.in file?

NitinAtIntel replied:
Right now the entrypoint script will replace this at run time to match # of vcpus on the system.  I think the conf.in file is still used per the internal implementation

NitinAtIntel replied:
To add to my response, the crypto reference stack currently has scripts to manage nginx execution, this script has config values passed to it as parameters and then behind the scenes nginx still uses a config file.

suresh-srinivas commented on containers/wordpress/wp_opt/https/Dockerfile:
```diff
@@ -0,0 +1,53 @@
+RUN apt-get update && \
+    apt-get install -y libtool
+
+# Build/Install IPP cyrypto library for the multi-buffer support
```
We may want to use a specific branch instead of master
IPP_CRYPTO_VERSION="ippcp_2020u3"

NitinAtIntel replied:
done

suresh-srinivas commented on containers/wordpress/wp_opt/https/Dockerfile:
```diff
@@ -0,0 +1,53 @@
+WORKDIR /home/${USERNAME}/temp
+RUN git clone  https://github.com/intel/QAT_Engine.git qat_engine && \
+    cd qat_engine && \
+    git checkout v0.6.1 && \
```
Suggest having versions for these instead of hardcoding.

NitinAtIntel replied:
done

suresh-srinivas commented on containers/wordpress/wp_opt/https/Dockerfile:
```diff
@@ -0,0 +1,53 @@
+    make && \
+    make install
+
+# Build/Install AYNCH NGINX
```
Again version so it is easy to update.

NitinAtIntel replied:
done

uttampawar commented:
@NitinAtIntel Any updated response to the reviewers comment?

NitinAtIntel commented:
> @NitinAtIntel Any updated response to the reviewers comment?

Due to some testing there was a delay. The main remaining item is removing hard coded versioning. Will not split into https base vs opt due to interdependencies, if we did so then one PR would have to be approved and then another and no possibility of reviewing them in true parallel.  Versioning update coming today and will ping for new review.

uttampawar commented on containers/wordpress/wp_opt/https/files/nginx.conf.in:
```diff
@@ -0,0 +1,125 @@
+worker_processes  96; # Depends on #vcpus
+
```
missing "include  qat_module.so" in the wp_opt.

NitinAtIntel replied:
done

uttampawar commented on containers/wordpress/wp_opt/https/files/nginx.conf.in:
```diff
@@ -0,0 +1,125 @@
+pid __NGINX_PID_FILE__;
+
+worker_rlimit_nofile 1000000;
+
```
missing "ssl_engine" block from the wp_opt in nginx.conf.in file.

NitinAtIntel replied:
done

uttampawar commented on containers/wordpress/wp_base/https/files/ssl-params.conf:
```diff
@@ -0,0 +1,12 @@
+add_header X-Frame-Options DENY;
+add_header X-Content-Type-Options nosniff;
+add_header X-XSS-Protection "1; mode=block";
+
```
This whole file is redundant in the latest version of PKB, we should do the same for simplicity.

NitinAtIntel replied:
Do you mean it is not required any more?

uttampawar reviewed (changes requested):
I see minor dependencies between wp_https_base and wp_https_opt version of docker file. My strong recommendation is to separate into totally independent two PRs. Is it too hard to have two separate docker files (which is just a list of steps) for https_base and https_opt? This way the OPT version can change as we get new support from any new components, over time from newer version of libraries, processor, etc.

NitinAtIntel commented:
@uttampawar Two docker files are already present, please see docker-compose.yml for the build order/dependency chain.

NitinAtIntel commented:
> This looks great. Thanks for starting this.
> 
> I would suggest using the same Nginx for both the base and the opt. And use a variable to enable/disable the OpenSSL crypto engine. See the following
> https://github.com/intel/crypto-reference-stack/blob/main/wordpress/centos8/Dockerfile
> 
> Would also suggest providing the NGINX_CIPHER as a variable.
> 
> And for us to be able to compare this to the wp_base and wp_opt, you can also have the NGINX_HTTP variable which controls the server type like in the reference stack.

nginx async is enabled on both base and opt now.  Regarding environment variables for nginx configuration, the crypto reference stack takes these variables into scripts that manage nginx execution. Behind the scenes, a config file is still used. If needed we can either bring those scripts in / clone them from the reference stack.

NitinAtIntel commented:
@uttampawar can you take another look, please? while there is a single PR (as one dockerfile depends on the other, included dockerfile), there are two docker files. 
@suresh-srinivas I either addressed all your comments or replied, happy to make further changes if needed.

suresh-srinivas commented on containers/wordpress/wp_base/https/files/nginx.conf.in:
```diff
@@ -0,0 +1,122 @@
+    ssl_session_timeout  300s;
+    ssl_prefer_server_ciphers   on;
+    ssl_protocols  TLSv1.2; #or TLSv1.3;
+
```
Are we supporting only TLS v1.2?

uttampawar replied:
+1 for only mention of TLSv1.2

suresh-srinivas reviewed (approved):
LGTM. Thanks for addressing all the feedback. I had one question which was around TLSv1.2 and TLSv1.3. The current file seems to supports only TLSv1.2?

uttampawar commented:
> > This looks great. Thanks for starting this.
> > I would suggest using the same Nginx for both the base and the opt. And use a variable to enable/disable the OpenSSL crypto engine. See the following
> > https://github.com/intel/crypto-reference-stack/blob/main/wordpress/centos8/Dockerfile
> > Would also suggest providing the NGINX_CIPHER as a variable.
> > And for us to be able to compare this to the wp_base and wp_opt, you can also have the NGINX_HTTP variable which controls the server type like in the reference stack.
> 
> nginx async is enabled on both base and opt now. Regarding environment variables for nginx configuration, the crypto reference stack takes these variables into scripts that manage nginx execution. Behind the scenes, a config file is still used. If needed we can either bring those scripts in / clone them from the reference stack.

Why do we have Intel version of Nginx with async in base run?
